core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala

```diff
@@ -210,28 +217,33 @@ trait DeltaSourceBase extends Source
       fromIndex: Long,
       isStartingVersion: Boolean,
       limits: Option[AdmissionLimits] = Some(new AdmissionLimits())):
-    ClosableIterator[IndexedFile] = {
-    if (options.readChangeFeed) {
-      // in this CDC use case, we need to consider RemoveFile and AddCDCFiles when getting the
+  ClosableIterator[IndexedFile] = {
+    var iter = if (options.readChangeFeed) {
+      // In this CDC use case, we need to consider RemoveFile and AddCDCFiles when getting the
       // offset.
 
       // This method is only used to get the offset so we need to return an iterator of IndexedFile.
       getFileChangesForCDC(fromVersion, fromIndex, isStartingVersion, limits, None).flatMap(_._2)
         .toClosable
     } else {
       val changes = getFileChanges(fromVersion, fromIndex, isStartingVersion)
-      if (limits.isEmpty) return changes
 
       // Take each change until we've seen the configured number of addFiles. Some changes don't
       // represent file additions; we retain them for offset tracking, but they don't count towards
       // the maxFilesPerTrigger conf.
-      val admissionControl = limits.get
-      changes.withClose { it =>
-        it.takeWhile { index =>
-          admissionControl.admit(Option(index.add))
+      if (limits.isEmpty) {
+        changes
+      } else {
+        val admissionControl = limits.get
+        changes.withClose { it =>
+          it.takeWhile { index =>
+            admissionControl.admit(Option(index.add))
+          }
         }
       }
     }
+    iter = stopIndexedFileIteratorAtSchemaChangeBarrier(iter)
```

Comment:
why not just return the result of `stopIndexedFileIteratorAtSchemaChangeBarrier(iter)`

Reply:
ah, good eye!